tools/psa/build_tfm.py

```diff
@@ -0,0 +1,118 @@
+MBEDTLS_GIT = 'https://github.com/ARMmbed/mbedtls.git'
+MBEDTLS_GIT_BRANCH = 'mbedtls-2.7.9'
+MBEDCRYPTO_GIT = 'https://github.com/ARMmbed/mbed-crypto.git'
+MBEDCRYPTO_GIT_BRANCH = 'mbedcrypto-1.1.0'
+CMSIS_GIT = 'https://github.com/ARM-software/CMSIS_5.git'
+CMSIS_GIT_BRANCH = '5.5.0'
+
+def run_cmd_return_output(command):
+    """
+    Run the command in the sytem and return output.
+    Commands are passed as a list of tokens.
+    E.g. The command 'git remote -v' would be passed in as:
+     ['git', 'remote', '-v']
+
+    :param command: System command as a list of tokens
+    """
+    output = ''
+    logger.debug('[Exec] %s', ' '.join(command))
+    try:
+        fnull = open(os.devnull, 'w')
+        output = subprocess.check_output(command, stderr=fnull)
+    except subprocess.CalledProcessError as e:
+        logger.error("The command %s failed with return code: %s",
+                          (' '.join(command)), e.returncode)
+        sys.exit(1)
+    return output
+
+def detect_write_tfm_version(tfm_dir, ):
+    """
+    Identify the version of TF-M and write it to VERSION.TXT
+    :param tfm_dir: The filesystem path where TF-M repo is cloned
+    """
+    cmd = ['git', '-C', tfm_dir, 'describe', '--tags', '--abbrev=12', '--dirty', '--always']
+    tfm_version = run_cmd_return_output(cmd)
+    logger.debug('TF-M version: %s', tfm_version)
+    with open(join(VERSION_FILE_PATH, 'VERSION.TXT'), 'w+') as f:
+        f.writelines(tfm_version.splitlines(True)[0])
+
+def check_clone_repo(basedir, name, git_branch, git):
+    """
+    Test if the repositories are already cloned. If not clone them
+    :param basedir: Base directory into which the repos will be cloned
+    :param name: Name of the git repository
+    :param git_branch: Git branch to be cloned
+    :param git: Web link to the git respository
+    """
+    if not isdir(join(basedir, name)):
+        logger.info('Cloning %s repo', name)
+        cmd = ['git', '-C', basedir, 'clone', '-b', git_branch, git]
+        _out = run_cmd_return_output(cmd)
+        logger.info('Cloned %s repo successfully', name)
+    else:
+        logger.info('%s repo exists', name)
+
+def clone_tfm_repo():
+    """
+    Clone TF-M git repos and it's dependencies
+    """
+    check_clone_repo(TF_M_BUILD_DIR, 'trusted-firmware-m', TFM_GIT_BRANCH, TFM_GIT)
+    check_clone_repo(TF_M_BUILD_DIR, 'mbedtls', MBEDTLS_GIT_BRANCH, MBEDTLS_GIT)
+    check_clone_repo(TF_M_BUILD_DIR, 'mbed-crypto', MBEDCRYPTO_GIT_BRANCH, MBEDCRYPTO_GIT)
+    check_clone_repo(TF_M_BUILD_DIR, 'CMSIS_5', CMSIS_GIT_BRANCH, CMSIS_GIT)
+
+    detect_write_tfm_version(join(TF_M_BUILD_DIR, 'trusted-firmware-m'), )
+
+def main():
+    """
+    Build Trusted Firmware M (TF-M) image for mbed-os supported TF-M targets.
+    Current version of the script only clones TF-M git repo and dependencies
+    and creates a VERSION.TXT file under 'components/TARGET_PSA/TARGET_TFM'
+    """
+    if not isdir(TF_M_BUILD_DIR):
```

Comment:
It would be ideal to delete all the files if the `TF_M_BUILD_DIR` is present. Because we are not checking which version of cloned repo is present when calling `check_clone_repo`

Reply:
@jainvikas8 It might be possible that users might have changes in the tf-m build directory. So it is not safe to assume that it can be deleted. Added a check to verify the version of cloned repos with the expected ones if the tf-m build directory already exists and warn the user if the versions don't match.

Reply:
Well, this would imply users can only have their changes saved locally, but can't pull in their git branches as the tag/version would change to a different HEAD. Is my understanding right?